libcontainer/nsenter/cloned_binary.c

```diff
@@ -192,49 +190,149 @@ static int fetchve(char ***argv)
 	return -EINVAL;
 }
 
+#define EFD_NONE     0x00
+#define EFD_MEMFD    0x01
+#define EFD_TMPFILE  0x02
+#define EFD_FILE     0x04
+#define EFD_ALL      (EFD_MEMFD | EFD_TMPFILE | EFD_FILE)
+
+static int make_execfd(int *fdtype, int allowed)
+{
+	int fd;
+	char template[] = "/tmp/runc-cloned-binary.XXXXXX";
+
+	/*
+	 * Try memfd first, it's much nicer since it's easily detected thanks to
+	 * sealing and also doesn't require assumptions like /tmp.
+	 */
+	*fdtype = EFD_MEMFD;
+	if (*fdtype & allowed) {
+		fd = memfd_create(RUNC_MEMFD_COMMENT, MFD_CLOEXEC | MFD_ALLOW_SEALING);
+		if (fd >= 0)
+			return fd;
+		if (errno != ENOSYS)
+			goto err;
+	}
+
+	/*
+	 * Try O_TMPFILE to avoid races where someone might snatch our file. Note
+	 * that O_EXCL isn't actually a security measure here (since you can just
+	 * fd re-open it and clear O_EXCL).
+	 */
+	*fdtype = EFD_TMPFILE;
+	if (*fdtype & allowed) {
+		fd = open("/tmp", O_TMPFILE | O_EXCL | O_RDWR | O_CLOEXEC, 0700);
+		if (fd >= 0) {
+			struct stat statbuf = {};
+			bool working_otmpfile = false;
+
+			/*
+			 * open(2) ignores unknown O_* flags -- yeah, I was surprised when I
+			 * found this out too. As a result we can't check for EINVAL. However,
+			 * if we get nlink != 0 (or EISDIR) then we know that this kernel
+			 * doesn't support O_TMPFILE.
+			 */
+			if (fstat(fd, &statbuf) >= 0)
+				working_otmpfile = (statbuf.st_nlink == 0);
+
+			if (working_otmpfile)
+				return fd;
+
+			/* Pretend that we got EISDIR since O_TMPFILE failed. */
+			close(fd);
+			errno = EISDIR;
+		}
+		if (errno != EISDIR)
+			goto err;
+	}
+
+	/*
+	 * Our final option is to create a temporary file the old-school way, and
+	 * then unlink it so that nothing else sees it by accident.
+	 */
+	*fdtype = EFD_FILE;
+	if (*fdtype & allowed) {
+		fd = mkostemp(template, O_CLOEXEC);
+		if (fd >= 0) {
+			if (!unlink(template))
+				return fd;
+			close(fd);
```

Comment:
You may want to consider:
```c
int _e_ = errno;
close(fd);
errno = _e_;
```

Reply:
I'm not sure we need it, we don't use errno here.